metricbeat/module/vsphere/host/host.go

```diff
@@ -111,9 +111,13 @@ func (m *MetricSet) Fetch() ([]common.MapStr, error) {
 	}
 
 	for _, hs := range hst {
-		totalCPU := int64(hs.Summary.Hardware.CpuMhz) * int64(hs.Summary.Hardware.NumCpuCores)
-		freeCPU := int64(totalCPU) - int64(hs.Summary.QuickStats.OverallCpuUsage)
-		freeMemory := int64(hs.Summary.Hardware.MemorySize) - (int64(hs.Summary.QuickStats.OverallMemoryUsage) * 1024 * 1024)
+		var totalCPU, freeCPU, freeMemory, memoryTotalBytes int64
+		if hs.Summary.Hardware != nil {
```

Comment:
What is the reason hs.Summary.Hardward is nil? Should we log an error or at least on Debug level that it is nil so we can investigate?

Reply:
`Hardware` is a pointer in `Summary` struct so if the library fails parsing the summary, it's zero value will be `nil`. In any casy any pointer, slice or interface access must always be preceded by a nil check. 

We could investigate further in the `gvmomi` library but maybe this behavior is intended and they set the value as `nil` when they cannot fetch enough data from the _vm_ to force a triple state on their data (0, X and `nil`)

I wouldn't log an error, at least not without further investigation because the error will only be useful for us (the developers) as it will be something like _Summary information was missing, this is potentially an parsing error on vsphere library, the vm not found or an unknown reason_ :smile:

Reply:
That sounds like a Debug log message could be useful here then. If a user asks why these values do not show up / are not populate we can ask to enable the debug logs to see such message. Could you add these?

As we have so far not seen many reports of this panics I assume it is some specific config users have to enable in vsphere to not have the info reported. Could be permissions? Unfortunately I don't know enough about vsphere.

Reply:
I really don't know either, I also don't have much experience. I guess that it could be a ton of causes because, all in all, you can use vsphere to manage an entire rack of hosts and vm.